panic, segmentation violation, and crash with Terraform 0.7.8

### Terraform Version

Terraform v0.7.8

### Affected Resource(s)

I can't tell what causes the crash. The Terraform code I'm running uses the following resources:

* aws_instance
* aws_iam_role
* aws_iam_instance_profile
* aws_security_group
* aws_security_group_rule
* aws_eip
* aws_route53_record
* aws_iam_policy
* aws_iam_policy_document
* aws_iam_policy_attachment
* terraform_remote_state
* template_file

### Terraform Configuration Files

It's a fairly large amount of code and I can't tell which part is causing the problem. The last log line suggests it had just finished creating the EIP for the EC2 Instance, so perhaps that's the problem. Here are the snippets for those two resources:

```hcl
resource "aws_instance" "instance" {
  ami = "${var.ami}"
  instance_type = "${var.instance_type}"
  iam_instance_profile = "${aws_iam_instance_profile.instance.name}"
  key_name = "${var.keypair_name}"
  vpc_security_group_ids = ["${aws_security_group.instance.id}"]
  subnet_id = "${var.subnet_id}"
  user_data = "${var.user_data}"
  tags {
    Name = "${var.name}"
  }
}

resource "aws_eip" "instance" {
  count = "${signum(var.attach_eip)}"

  instance = "${aws_instance.instance.id}"
  vpc = true

  # https://github.com/hashicorp/terraform/issues/1815
  provisioner "local-exec" {
    command = "echo 'Sleeping 15 seconds to work around EIP propagation bug in Terraform' && sleep 15"
  }
}
```

Perhaps it's something to do with using `count` or `local-exec` in the `aws_eip` resource?

### Panic Output

The full crash log is here: https://gist.github.com/brikis98/634699f731539a04e1ff7ec0e8b54d0b

```
module.bastion.aws_eip.instance: Provisioning with 'local-exec'...
module.bastion.aws_eip.instance (local-exec): Executing: /bin/sh -c "echo 'Sleeping 15 seconds to work around EIP propagation bug in Terraform' && sleep 15"
module.bastion.aws_eip.instance (local-exec): Sleeping 15 seconds to work around EIP propagation bug in Terraform
module.bastion.aws_eip.instance: Still creating... (10s elapsed)
module.bastion.aws_eip.instance: Creation complete
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xf1f24]

goroutine 1609 [running]:
panic(0x25e3540, 0xc42000e090)
	/opt/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/terraform/terraform.(*EvalApplyProvisioners).apply(0xc420505170, 0x431cc20, 0xc4208b2a90, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_apply.go:278 +0x5b4
github.com/hashicorp/terraform/terraform.(*EvalApplyProvisioners).Eval(0xc420505170, 0x431cc20, 0xc4208b2a90, 0x2, 0x2, 0xc420261600, 0xc)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_apply.go:179 +0x151
github.com/hashicorp/terraform/terraform.EvalRaw(0x4302480, 0xc420505170, 0x431cc20, 0xc4208b2a90, 0x29882a0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc42025d6e0, 0x431cc20, 0xc4208b2a90, 0x2, 0x2, 0xc420279130, 0xc)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0x6c
github.com/hashicorp/terraform/terraform.EvalRaw(0x4302d80, 0xc42025d6e0, 0x431cc20, 0xc4208b2a90, 0x4302d80, 0xc42025d6e0, 0xc420505260, 0x24)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.Eval(0x4302d80, 0xc42025d6e0, 0x431cc20, 0xc4208b2a90, 0xc42025d6e0, 0x4302d80, 0xc42025d6e0, 0xc420965b30)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x53
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2a939e0, 0xc4203b8128, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:249 +0xa0f
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc4205cf240, 0xc4207be280, 0xc4205cf250, 0xc4207d0de0, 0xc4205cf260, 0x2a939e0, 0xc4203b8128, 0xc42078bda0, 0xc42089cd20)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/dag.go:238 +0x251
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
	/opt/gopath/src/github.com/hashicorp/terraform/dag/dag.go:247 +0x618
```

### Expected Behavior

`terraform apply` should have succeeded.

### Actual Behavior

It crashed.

### Steps to Reproduce

Run `terraform apply`

### Important Factoids

This code worked fine before upgrading to Terraform 0.7.8

Same for us as well. `terraform plan` on 0.7.8 with a non-trivial config works fine. `terraform apply` on the same plan segfaults.

Works as expected on 0.7.5. Haven't had a chance with .6 or .7 yet.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xf1f24]

goroutine 8019 [running]:
panic(0x25e3540, 0xc4200120a0)
	/opt/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/terraform/terraform.(*EvalApplyProvisioners).apply(0xc4215d9230, 0x431cc20, 0xc4200725b0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_apply.go:278 +0x5b4
github.com/hashicorp/terraform/terraform.(*EvalApplyProvisioners).Eval(0xc4215d9230, 0x431cc20, 0xc4200725b0, 0x2, 0x2, 0xc420ee5d90, 0x9)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_apply.go:179 +0x151
github.com/hashicorp/terraform/terraform.EvalRaw(0x4302480, 0xc4215d9230, 0x431cc20, 0xc4200725b0, 0x29882a0, 0x0, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.(*EvalSequence).Eval(0xc4206ae5e0, 0x431cc20, 0xc4200725b0, 0x2, 0x2, 0xc420ee5a20, 0x9)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval_sequence.go:10 +0x6c
github.com/hashicorp/terraform/terraform.EvalRaw(0x4302d80, 0xc4206ae5e0, 0x431cc20, 0xc4200725b0, 0x4302d80, 0xc4206ae5e0, 0xc4215d9320, 0x26)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:53 +0x17f
github.com/hashicorp/terraform/terraform.Eval(0x4302d80, 0xc4206ae5e0, 0x431cc20, 0xc4200725b0, 0xc4206ae5e0, 0x4302d80, 0xc4206ae5e0, 0xc41ff54e44)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/eval.go:34 +0x53
github.com/hashicorp/terraform/terraform.(*Graph).walk.func1(0x2a939e0, 0xc4204ba260, 0x0, 0x0)
	/opt/gopath/src/github.com/hashicorp/terraform/terraform/graph.go:249 +0xa0f
github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk.func3(0xc4215768b0, 0xc4209a2840, 0xc4215768c0, 0xc421431890, 0xc4215768d0, 0x2a939e0, 0xc4204ba260, 0xc421392960, 0xc42132c420)
	/opt/gopath/src/github.com/hashicorp/terraform/dag/dag.go:238 +0x251
created by github.com/hashicorp/terraform/dag.(*AcyclicGraph).Walk
	/opt/gopath/src/github.com/hashicorp/terraform/dag/dag.go:247 +0x618
```

It's a 14M crash log if anyone's interested in seeing the whole thing.

--- a/terraform/context_apply_test.go
+++ b/terraform/context_apply_test.go
@@ -2246,6 +2246,43 @@ func TestContext2Apply_providerConfigureDisabled(t *testing.T) {
 	}
 }
 
+func TestContext2Apply_provisionerModule(t *testing.T) {
+	m := testModule(t, "apply-provisioner-module")
+	p := testProvider("aws")
+	pr := testProvisioner()
+	p.ApplyFn = testApplyFn
+	p.DiffFn = testDiffFn
+	ctx := testContext2(t, &ContextOpts{
+		Module: m,
+		Providers: map[string]ResourceProviderFactory{
+			"aws": testProviderFuncFixed(p),
+		},
+		Provisioners: map[string]ResourceProvisionerFactory{
+			"shell": testProvisionerFuncFixed(pr),
+		},
+	})
+
+	if _, err := ctx.Plan(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	state, err := ctx.Apply()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	actual := strings.TrimSpace(state.String())
+	expected := strings.TrimSpace(testTerraformApplyProvisionerModuleStr)
+	if actual != expected {
+		t.Fatalf("bad: \n%s", actual)
+	}
+
+	// Verify apply was invoked
+	if !pr.ApplyCalled {
+		t.Fatalf("provisioner not invoked")
+	}
+}
+
 func TestContext2Apply_Provisioner_compute(t *testing.T) {
 	m := testModule(t, "apply-provisioner-compute")
 	p := testProvider("aws")

--- a/terraform/graph_builder_apply_test.go
+++ b/terraform/graph_builder_apply_test.go
@@ -100,16 +100,16 @@ meta.count-boundary (count boundary fixup)
   module.child.aws_instance.create
   module.child.aws_instance.other
   module.child.provider.aws
+  module.child.provisioner.exec
   provider.aws
-  provisioner.exec
 module.child.aws_instance.create
   module.child.provider.aws
-  provisioner.exec
+  module.child.provisioner.exec
 module.child.aws_instance.other
   module.child.aws_instance.create
   module.child.provider.aws
 module.child.provider.aws
   provider.aws
+module.child.provisioner.exec
 provider.aws
-provisioner.exec
 `

--- a/terraform/terraform_test.go
+++ b/terraform/terraform_test.go
@@ -522,6 +522,13 @@ aws_instance.foo:
   type = aws_instance
 `
 
+const testTerraformApplyProvisionerModuleStr = `
+<no state>
+module.child:
+  aws_instance.bar:
+    ID = foo
+`
+
 const testTerraformApplyProvisionerFailStr = `
 aws_instance.bar: (tainted)
   ID = foo

--- a/terraform/test-fixtures/apply-provisioner-module/child/main.tf
+++ b/terraform/test-fixtures/apply-provisioner-module/child/main.tf
@@ -1,0 +1,5 @@
+resource "aws_instance" "bar" {
+    provisioner "shell" {
+        foo = "bar"
+    }
+}

--- a/terraform/test-fixtures/apply-provisioner-module/main.tf
+++ b/terraform/test-fixtures/apply-provisioner-module/main.tf
@@ -1,0 +1,1 @@
+module "child" { source = "./child" }

--- a/terraform/test-fixtures/transform-provisioner-module/child/main.tf
+++ b/terraform/test-fixtures/transform-provisioner-module/child/main.tf
@@ -1,0 +1,3 @@
+resource "aws_instance" "foo" {
+    provisioner "shell" {}
+}

--- a/terraform/test-fixtures/transform-provisioner-module/main.tf
+++ b/terraform/test-fixtures/transform-provisioner-module/main.tf
@@ -1,0 +1,7 @@
+resource "aws_instance" "foo" {
+  provisioner "shell" {}
+}
+
+module "child" {
+  source = "./child"
+}

--- a/terraform/transform_provisioner.go
+++ b/terraform/transform_provisioner.go
@@ -40,14 +40,15 @@ func (t *ProvisionerTransformer) Transform(g *Graph) error {
 	for _, v := range g.Vertices() {
 		if pv, ok := v.(GraphNodeProvisionerConsumer); ok {
 			for _, p := range pv.ProvisionedBy() {
-				if m[p] == nil {
+				key := provisionerMapKey(p, pv)
+				if m[key] == nil {
 					err = multierror.Append(err, fmt.Errorf(
 						"%s: provisioner %s couldn't be found",
 						dag.VertexName(v), p))
 					continue
 				}
 
-				g.Connect(dag.BasicEdge(v, m[p]))
+				g.Connect(dag.BasicEdge(v, m[key]))
 			}
 		}
 	}
@@ -80,8 +81,21 @@ func (t *MissingProvisionerTransformer) Transform(g *Graph) error {
 			continue
 		}
 
+		// If this node has a subpath, then we use that as a prefix
+		// into our map to check for an existing provider.
+		var path []string
+		if sp, ok := pv.(GraphNodeSubPath); ok {
+			raw := normalizeModulePath(sp.Path())
+			if len(raw) > len(rootModulePath) {
+				path = raw
+			}
+		}
+
 		for _, p := range pv.ProvisionedBy() {
-			if _, ok := m[p]; ok {
+			// Build the key for storing in the map
+			key := provisionerMapKey(p, pv)
+
+			if _, ok := m[key]; ok {
 				// This provisioner already exists as a configure node
 				continue
 			}
@@ -92,8 +106,23 @@ func (t *MissingProvisionerTransformer) Transform(g *Graph) error {
 				continue
 			}
 
+			// Build the vertex
+			var newV dag.Vertex = &graphNodeProvisioner{ProvisionerNameValue: p}
+			if len(path) > 0 {
+				// If we have a path, we do the flattening immediately. This
+				// is to support new-style graph nodes that are already
+				// flattened.
+				if fn, ok := newV.(GraphNodeFlattenable); ok {
+					var err error
+					newV, err = fn.Flatten(path)
+					if err != nil {
+						return err
+					}
+				}
+			}
+
 			// Add the missing provisioner node to the graph
-			m[p] = g.Add(&graphNodeProvisioner{ProvisionerNameValue: p})
+			m[key] = g.Add(newV)
 		}
 	}
 
@@ -129,6 +158,20 @@ func (t *CloseProvisionerTransformer) Transform(g *Graph) error {
 	}
 
 	return nil
+}
+
+// provisionerMapKey is a helper that gives us the key to use for the
+// maps returned by things such as provisionerVertexMap.
+func provisionerMapKey(k string, v dag.Vertex) string {
+	pathPrefix := ""
+	if sp, ok := v.(GraphNodeSubPath); ok {
+		raw := normalizeModulePath(sp.Path())
+		if len(raw) > len(rootModulePath) {
+			pathPrefix = modulePrefixStr(raw) + "."
+		}
+	}
+
+	return pathPrefix + k
 }
 
 func provisionerVertexMap(g *Graph) map[string]dag.Vertex {

--- a/terraform/transform_provisioner_test.go
+++ b/terraform/transform_provisioner_test.go
@@ -39,6 +39,74 @@ func TestMissingProvisionerTransformer(t *testing.T) {
 	}
 }
 
+func TestMissingProvisionerTransformer_module(t *testing.T) {
+	mod := testModule(t, "transform-provisioner-module")
+
+	g := Graph{Path: RootModulePath}
+	{
+		concreteResource := func(a *NodeAbstractResource) dag.Vertex {
+			return a
+		}
+
+		var state State
+		state.init()
+		state.Modules = []*ModuleState{
+			&ModuleState{
+				Path: []string{"root"},
+				Resources: map[string]*ResourceState{
+					"aws_instance.foo": &ResourceState{
+						Primary: &InstanceState{ID: "foo"},
+					},
+				},
+			},
+
+			&ModuleState{
+				Path: []string{"root", "child"},
+				Resources: map[string]*ResourceState{
+					"aws_instance.foo": &ResourceState{
+						Primary: &InstanceState{ID: "foo"},
+					},
+				},
+			},
+		}
+
+		tf := &StateTransformer{
+			Concrete: concreteResource,
+			State:    &state,
+		}
+		if err := tf.Transform(&g); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+	}
+
+	{
+		transform := &AttachResourceConfigTransformer{Module: mod}
+		if err := transform.Transform(&g); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+	}
+
+	{
+		transform := &MissingProvisionerTransformer{Provisioners: []string{"shell"}}
+		if err := transform.Transform(&g); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+	}
+
+	{
+		transform := &ProvisionerTransformer{}
+		if err := transform.Transform(&g); err != nil {
+			t.Fatalf("err: %s", err)
+		}
+	}
+
+	actual := strings.TrimSpace(g.String())
+	expected := strings.TrimSpace(testTransformMissingProvisionerModuleStr)
+	if actual != expected {
+		t.Fatalf("bad:\n\n%s", actual)
+	}
+}
+
 func TestCloseProvisionerTransformer(t *testing.T) {
 	mod := testModule(t, "transform-provisioner-basic")
 
@@ -93,6 +161,15 @@ func TestGraphNodeProvisioner_ProvisionerName(t *testing.T) {
 const testTransformMissingProvisionerBasicStr = `
 aws_instance.web
   provisioner.shell
+provisioner.shell
+`
+
+const testTransformMissingProvisionerModuleStr = `
+aws_instance.foo
+  provisioner.shell
+module.child.aws_instance.foo
+  module.child.provisioner.shell
+module.child.provisioner.shell
 provisioner.shell
 `
 

